src/storage/StorageServer.cpp

```diff
@@ -0,0 +1,166 @@
+    workers_ = apache::thrift::concurrency::PriorityThreadManager::newPriorityThreadManager(
+                                 FLAGS_num_worker_threads, true /*stats*/);
+    workers_->setNamePrefix("executor");
+    workers_->start();
+
+    std::string clusteridFile =
+        folly::stringPrintf("%s/%s", dataPaths_[0].c_str(), "/storage.cluster.id");
+    clusterMan_
+        = std::make_unique<nebula::meta::ClusterManager>("", clusteridFile);
+    if (!clusterMan_->loadClusterId()) {
+        LOG(INFO) << "storaged misses clusterId";
+    }
+    // Meta client
+    metaClient_ = std::make_unique<meta::MetaClient>(ioThreadPool_,
+                                                     metaAddrs_,
+                                                     localHost_,
+                                                     clusterMan_.get(),
+                                                     true);
+    if (!metaClient_->waitForMetadReady()) {
+        LOG(ERROR) << "waitForMetadReady error!";
+        return false;
+    }
+
+    gFlagsMan_ = std::make_unique<meta::ClientBasedGflagsManager>(metaClient_.get());
+    gFlagsMan_->init();
+
+    LOG(INFO) << "Init schema manager";
+    schemaMan_ = meta::SchemaManager::create();
+    schemaMan_->init(metaClient_.get());
+
+    LOG(INFO) << "Init kvstore";
+    kvstore_ = getStoreInstance();
+
+    if (nullptr == kvstore_) {
+        LOG(ERROR) << "Init kvstore failed";
+        return false;
+    }
+
+    if (!initWebService()) {
+        LOG(ERROR) << "Init webservice failed!";
+        return false;
+    }
+
+    auto handler = std::make_shared<StorageServiceHandler>(kvstore_.get(), schemaMan_.get());
+    try {
+        LOG(INFO) << "The storage deamon start on " << localHost_;
+        tfServer_ = std::make_unique<apache::thrift::ThriftServer>();
+        tfServer_->setPort(FLAGS_port);
+        tfServer_->setReusePort(FLAGS_reuse_port);
+        tfServer_->setIdleTimeout(std::chrono::seconds(0));  // No idle timeout on client connection
+        tfServer_->setIOThreadPool(ioThreadPool_);
+        tfServer_->setThreadManager(workers_);
+        tfServer_->setInterface(std::move(handler));
+        tfServer_->setStopWorkersOnStopListening(false);
+        tfServer_->serve();  // Will wait until the server shuts down
+    } catch (const std::exception& e) {
+        LOG(ERROR) << "Start thrift server failed, error:" << e.what();
+        return false;
+    }
+    return true;
+}
+
+void StorageServer::stop() {
+    if (stopped_) {
+        LOG(INFO) << "All services has been stopped";
+        return;
+    }
+    if (webStatus_ == Status::RUNNING) {
+        nebula::WebService::stop();
+        webStatus_ = Status::STOPPED;
+    }
+    if (metaClient_) {
+        metaClient_->stop();
+    }
+    if (kvstore_) {
+        kvstore_.reset();
+    }
+    if (tfServer_) {
+        tfServer_->stop();
+    }
+    stopped_ = true;
```

Comment:
move to line 148, maybe need change to atomic_bool, else set true here is useless

Reply:
what's the meaning?  stopped_ has been atomic_bool yet.

Reply:
the stopped_ look like to stop multi thread to entry this function, but it useless if set the value at the last, maybe we should use exchange.

Reply:
the stopped_ look like to stop multi thread to entry this function, but it useless if set the value at the last, maybe we should use exchange.

Reply:
@wadeliuyi Not exactly.  stopped_ here is used to protect the server from being stopped twice.

Reply:
I am not sure the case when user send a SIGTERM signal to storage service, the storage service will process this signal in some thread context, maybe it is not main context. then the main context exit from gServer->start(), and call gServer's dctor, maybe this is a multi thread env.

Reply:
I got your point.  Yes,  let's move stopped_ = true before thrift stop